### PR TITLE
feat: campaign introduction settings panel 

### LINKED ISSUE #18: feat: campaign introduction settings panel
# Feature Request

## User Story

As an admin, I want manage campaign introduction settings so that I can easily and visually create a campaign.

## Visuals

https://drive.google.com/open?id=1lsI_WRmOiKiS0VIcxmb5f-rgRA2VCVGQ

## Possible Solution

Campaign introduction settings should include: 
- Layout (Order of heading/text/image)
- Heading text
- Subheading text
- Image upload
- Next button text

## Related

feat: campaign editor settings sidebar #17 

## Acceptance Criteria

- [x] Panel contains the settings listed above
- [x] Panel can be included in the settings sidebar
- [x] Panel settings changes are reflected in the preview
- [x] Use WP Media popup to upload and select a featured image for a campaign

### PR DESCRIPTION
**Pull request by ravinderk**
## Description
This PR will resolve #18 

## How Has This Been Tested?
I tested this PR by choosing a setting from available options.

## Screenshots (jpeg or gifs if applicable):
image

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: resources/js/app/campaign-editor/components/campaign-preview/index.jsx [hunk 1]
```diff
@@ -41,7 +42,9 @@ const campaignDesign = {
 	maxWidth: '430px',
 };
 
-function CampaignPreview( { campaignPage } ) {
+function CampaignPreview( { campaignPage, settings } ) {
```
Comment (JasonTheAdams) [inline resources/js/app/campaign-editor/components/campaign-preview/index.jsx]: Break out the various views into separate components as this component is huge.

File: resources/js/app/campaign-editor/components/settings-panel/views/Campaigns/config.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+	return pages;
+}
+
+function getSettingsComponent() {
+	const settings = {};
+	config.map( ( { id, setting } ) => ( Object.assign( settings, { [ id ]: setting } ) ) );
+
+	return settings;
+}
```
Comment (JasonTheAdams) [inline resources/js/app/campaign-editor/components/settings-panel/views/Campaigns/config.js]: Please move the `const config` above these. It took me a moment to figure out where the `config` you're referencing within the functions is defined. I know this works due to hoisting, but it's counterintuitive.

File: resources/js/app/campaign-editor/components/settings-panel/views/Campaigns/views/CampaignsIntroduction.jsx [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	campaignPage: PropTypes.string.isRequired,
+};
+
+function mapStateToProps( { campaign = {}, campaignPage } ) {
+	return {
+		campaign,
+		campaignPage,
+	};
+}
+export default connect( mapStateToProps )( CampaignsIntroduction );
```
Comment (JasonTheAdams) [inline resources/js/app/campaign-editor/components/settings-panel/views/Campaigns/views/CampaignsIntroduction.jsx]: I've seen you do this before, where you define a component and then immediately connect it on export. This defeats the purpose of the [presentational and container component methodology](https://redux.js.org/basics/usage-with-react#implementing-components) as defined in Redux. A presentational component should not contain any application functionality such as `dispatch`, and a container component should not contain any markup. This component is doing both.

File: resources/js/app/campaign-editor/components/campaign-preview/index.jsx (hunk 1 above)
Comment (ravinderk) [reply resources/js/app/campaign-editor/components/campaign-preview/index.jsx]: I updated the logic of this component, no it does not need `settings`

File: resources/js/app/campaign-editor/components/settings-panel/views/Campaigns/views/CampaignsIntroduction.jsx (hunk 3 above)
Comment (ravinderk) [reply resources/js/app/campaign-editor/components/settings-panel/views/Campaigns/views/CampaignsIntroduction.jsx]: @JasonTheAdams You are right there. I tried to follow your suggestion.

To fix this issue, I created a container for `CampaignsIntroductionSettings`

File: resources/js/app/campaign-editor/campaign-types/general-fundraising/containers/CampaignIntroductionSettingsContainer.jsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import React from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import CampaignIntroductionSettings from '../settings/CampaignIntroductionSettings';
+import { handlerAddCampaignIntroductionSettings } from '../../../actions/campaign';
+import { getFieldName } from '../../../utils/fields';
```
Comment (ravinderk) [inline resources/js/app/campaign-editor/campaign-types/general-fundraising/containers/CampaignIntroductionSettingsContainer.jsx]: @JasonTheAdams @nathanielks I created GitHub issue to make helper function easy to use: https://github.com/impress-org/givewp-campaigns/issues/83

**Pull request merged**